# Module 1

   ### Welcome to the Answer notebook for Module 1 ! 
These notebooks have been provided to code and solve all the queries asked in the modules.

This environment has all the necessary libraries pre-installed, and all the Stock, Commodities and Index data files available at the following location - 



#### The problem statements and their corresponding answers are expected to be in the following format - 



#--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------#

### Query 1.1 
Import the csv file of the stock of your choosing using 'pd.read_csv()' function into a dataframe.
Shares of a company can be offered in more than one category. The category of a stock is indicated in the ‘Series’ column. If the csv file has data on more than one category, the ‘Date’ column will have repeating values. To avoid repetitions in the date, remove all the rows where 'Series' column is NOT 'EQ'.
Analyze and understand each column properly.
You'd find the head(), tail() and describe() functions to be immensely useful for exploration. You're free to carry out any other exploration of your own.

In [1]:
#The solution code should start right after the query statement, for example -
import numpy as np 
import pandas as pd
import warnings
warnings.filterwarnings('ignore')
#And solve the query

In [2]:
df = pd.read_csv('MINDTREE.csv')
df = df[df.Series == 'EQ']
df.head(5)

Symbol Series  ... Deliverable Qty  % Dly Qt to Traded Qty
0  MINDTREE     EQ  ...          100362                   36.90
1  MINDTREE     EQ  ...           82090                   39.65
2  MINDTREE     EQ  ...          150098                   38.35
3  MINDTREE     EQ  ...          170919                   20.05
4  MINDTREE     EQ  ...           57853                   16.42

[5 rows x 15 columns]

In [3]:
df.tail(5)

Symbol Series  ... Deliverable Qty  % Dly Qt to Traded Qty
489  MINDTREE     EQ  ...          884988                   71.20
490  MINDTREE     EQ  ...         3851210                   89.54
491  MINDTREE     EQ  ...         3479880                   91.89
492  MINDTREE     EQ  ...           26734                   12.42
493  MINDTREE     EQ  ...          222691                   45.83

[5 rows x 15 columns]

### Query 1.2
Calculate the maximum, minimum and mean price for the last 90 days. (price=Closing Price unless stated otherwise)

In [4]:
df_tail_closing_price = df['Close Price'].tail(90)
df_tail_closing_price.describe()

count     90.000000
mean     918.612222
std       46.589768
min      812.250000
25%      890.725000
50%      917.725000
75%      956.212500
max      989.400000
Name: Close Price, dtype: float64

Answers 1.2

1.   Maximum = 989.4
2.   Minimum = 812.25
3.   Mean    = 918.612222


**This is the expected format of the answer notebook**

### Query 1.3
Analyse the data types for each column of the dataframe. Pandas knows how to deal dates in an intelligent manner. But to make use of pandas funcationality for dates, you need to ensure that the columnis of type 'datetime64(ns)'. Change the date column from 'object' type to 'datetime64(ns)' for future convinience. See what happens if you substract the minimum value of the date column from the maximum value.  

In [5]:
# to get datatypes of every columns
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 494 entries, 0 to 493
Data columns (total 15 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   Symbol                  494 non-null    object 
 1   Series                  494 non-null    object 
 2   Date                    494 non-null    object 
 3   Prev Close              494 non-null    float64
 4   Open Price              494 non-null    float64
 5   High Price              494 non-null    float64
 6   Low Price               494 non-null    float64
 7   Last Price              494 non-null    float64
 8   Close Price             494 non-null    float64
 9   Average Price           494 non-null    float64
 10  Total Traded Quantity   494 non-null    int64  
 11  Turnover                494 non-null    float64
 12  No. of Trades           494 non-null    int64  
 13  Deliverable Qty         494 non-null    int64  
 14  % Dly Qt to Traded Qty  494 non-null    fl

In [6]:
# date has datatyepe 'object' 
# convert to 'datetime64(ns)'
df['Date'] = pd.to_datetime(df['Date'])
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 494 entries, 0 to 493
Data columns (total 15 columns):
 #   Column                  Non-Null Count  Dtype         
---  ------                  --------------  -----         
 0   Symbol                  494 non-null    object        
 1   Series                  494 non-null    object        
 2   Date                    494 non-null    datetime64[ns]
 3   Prev Close              494 non-null    float64       
 4   Open Price              494 non-null    float64       
 5   High Price              494 non-null    float64       
 6   Low Price               494 non-null    float64       
 7   Last Price              494 non-null    float64       
 8   Close Price             494 non-null    float64       
 9   Average Price           494 non-null    float64       
 10  Total Traded Quantity   494 non-null    int64         
 11  Turnover                494 non-null    float64       
 12  No. of Trades           494 non-null    int64     

In [7]:
max_date = max(df['Date'])
min_date = min(df['Date'])
result = max_date - min_date

In [8]:
# Result
print(f'Max Date = {max_date}, Min Date = {min_date}, \n result = {result}')

Max Date = 2019-05-13 00:00:00, Min Date = 2017-05-15 00:00:00, 
 result = 728 days 00:00:00


### Query 1.4
In a separate array , calculate the monthwise VWAP (Volume Weighted Average Price ) of the stock. ( VWAP = sum(price*volume)/sum(volume) ) To know more about VWAP , visit - VWAP definition {Hint : Create a new dataframe column ‘Month’. The values for this column can be derived from the ‘Date” column by using appropriate pandas functions. Similarly, create a column ‘Year’ and initialize it. Then use the 'groupby()' function by month and year. Finally, calculate the vwap value for each month (i.e. for each group created).

In [14]:
df['Month'] = pd.DatetimeIndex(df['Date']).month
df['Year'] = pd.DatetimeIndex(df['Date']).year

In [16]:
# Calculating VWAP
df['VWAP'] = np.cumsum(df['Close Price'] * df['Total Traded Quantity']) / np.cumsum(df['Total Traded Quantity'])


In [19]:
#Grouping
df_vwap = df[['Month', 'Year', 'VWAP']]
group = df_vwap.groupby(['Month', 'Year'])
group.first()

VWAP
Month Year            
1     2018  516.202743
      2019  843.125130
2     2018  572.722380
      2019  846.155182
3     2018  609.866603
      2019  848.327396
4     2018  651.524063
      2019  853.178143
5     2017  509.450000
      2018  730.334478
      2019  856.030810
6     2017  522.517683
      2018  760.847401
7     2017  528.654830
      2018  776.898919
8     2017  521.674873
      2018  806.405595
9     2017  514.842649
      2018  822.588413
10    2017  503.871258
      2018  840.417469
11    2017  501.226909
      2018  842.242954
12    2017  504.364596
      2018  842.393059

### Query 1.5
Write a function to calculate the average price over the last N days of the stock price data where N is a user defined parameter. Write a second function to calculate the profit/loss percentage over the last N days. Calculate the average price AND the profit/loss percentages over the course of last - 1 week, 2 weeks, 1 month, 3 months, 6 months and 1 year. {Note : Profit/Loss percentage between N days is the percentage change between the closing prices of the 2 days }

In [34]:
def avg_last_N_days(N):
  '''
  N = Number of days 
  '''

  return np.sum(pd.Series(df['Average Price'].tail(N))) / N

In [44]:
def prof_loss(N):
  '''
  N = Number of days 
  '''
  diff = df['Close Price'].tail(N).iloc[N-1] - df['Close Price'].tail(N).iloc[0]
  if diff < 0:
    loss_per = -(diff / (df['Close Price'].tail(N).iloc[N-1])) * 100
    return f'{loss_per}% loss'
  else:
    prof_per = (diff / (df['Close Price'].tail(N).iloc[N-1])) * 100
    prof_per
    return f'{prof_per}% profit'

In [45]:
# required outputs
print('Average prices for last N days are as follows: ')
print('For last 1 Week: ', avg_last_N_days(5))
print('For last 2 Weeks: ', avg_last_N_days(10))
print('For last 1 Month: ', avg_last_N_days(20))
print('For last 3 Months: ', avg_last_N_days(60))
print('For last 6 Months: ', avg_last_N_days(120))
print('For last 1 Year: ', avg_last_N_days(240))
print('*'*100)
print('Profit/Loss Percentages for last N days are as follows: ')
print('For last 1 Week: ', prof_loss(5))
print('For last 2 Weeks: ', prof_loss(10))
print('For last 1 Month: ', prof_loss(20))
print('For last 3 Months: ', prof_loss(60))
print('For last 6 Months: ', prof_loss(120))
print('For last 1 Year: ', prof_loss(240))

Average prices for last N days are as follows: 
For last 1 Week:  979.4680000000001
For last 2 Weeks:  979.175
For last 1 Month:  978.2695
For last 3 Months:  943.5338333333333
For last 6 Months:  903.1566666666666
For last 1 Year:  948.6203333333334
****************************************************************************************************
Profit/Loss Percentages for last N days are as follows: 
For last 1 Week:  0.0868144214074173% loss
For last 2 Weeks:  0.33193749361658664% profit
For last 1 Month:  0.9038913287713229% profit
For last 3 Months:  6.235318149320811% profit
For last 6 Months:  14.166070881421719% profit
For last 1 Year:  4.064957614135426% loss


### Query 1.6
Add a column 'Day_Perc_Change' where the values are the daily change in percentages i.e. the percentage change between 2 consecutive day's closing prices. Instead of using the basic mathematical formula for computing the same, use 'pct_change()' function provided by Pandas for dataframes. You will note that the first entry of the column will have a ‘Nan’ value. Why does this happen? Either remove the first row, or set the entry to 0 before proceeding.

In [46]:
df['Day_Perc_Change'] = df['Close Price'].pct_change().fillna(0)
df

Symbol Series       Date  ...  Month  Year  Day_Perc_Change
0    MINDTREE     EQ 2017-05-15  ...      5  2017         0.000000
1    MINDTREE     EQ 2017-05-16  ...      5  2017         0.003631
2    MINDTREE     EQ 2017-05-17  ...      5  2017        -0.009681
3    MINDTREE     EQ 2017-05-18  ...      5  2017        -0.026563
4    MINDTREE     EQ 2017-05-19  ...      5  2017         0.009333
..        ...    ...        ...  ...    ...   ...              ...
489  MINDTREE     EQ 2019-05-07  ...      5  2019        -0.001833
490  MINDTREE     EQ 2019-05-08  ...      5  2019         0.000561
491  MINDTREE     EQ 2019-05-09  ...      5  2019        -0.000051
492  MINDTREE     EQ 2019-05-10  ...      5  2019        -0.005202
493  MINDTREE     EQ 2019-05-13  ...      5  2019         0.003845

[494 rows x 19 columns]

### Query 1.7
Add another column 'Trend' whose values are: 

*   'Slight or No change' for 'Day_Perc_Change' in between -0.5 and 0.5  
*   'Slight positive' for 'Day_Perc_Change' in between 0.5 and 1
*   'Slight negative' for 'Day_Perc_Change' in between -0.5 and -1
*   'Positive' for 'Day_Perc_Change' in between 1 and 3
*   'Negative' for 'Day_Perc_Change' in between -1 and -3
*   'Among top gainers' for 'Day_Perc_Change' in between 3 and 7
*   'Among top losers' for 'Day_Perc_Change' in between -3 and -7
*   'Bull run' for 'Day_Perc_Change' >7
*   'Bear drop' for 'Day_Perc_Change' <-7
 

In [54]:
if ((df['Day_Perc_Change'] >= -0.5) & (df['Day_Perc_Change'] <= 0.5)).all():
    df['Trend'] = 'Slight or No change'
if ((df['Day_Perc_Change'] >= 0.5) & (df['Day_Perc_Change'] <= 1)).all():
    df['Trend'] = 'Slight positive'
if ((df['Day_Perc_Change'] <= -0.5) & (df['Day_Perc_Change'] >= -1)).all():
    df['Trend'] = 'Slight negative'
if ((df['Day_Perc_Change'] >= 1) & (df['Day_Perc_Change'] <= 3)).all():
    df['Trend'] = 'Positive' 
if ((df['Day_Perc_Change'] <= -1) & (df['Day_Perc_Change'] >= -3)).all():
    df['Trend'] = 'Negative'
if ((df['Day_Perc_Change'] >= 3) & (df['Day_Perc_Change'] <= 7)).all():
    df['Trend'] = 'Among top gainers'
if ((df['Day_Perc_Change'] <= -3) & (df['Day_Perc_Change'] >= -7)).all():
    df['Trend'] = 'Among top losers'
if (df['Day_Perc_Change'] > 7).all():
    df['Trend'] = 'Bull run' 
if (df['Day_Perc_Change'] < -7).all():
    df['Trend'] = 'Bear drop' 

In [56]:
df

Symbol Series       Date  ...  Year  Day_Perc_Change                Trend
0    MINDTREE     EQ 2017-05-15  ...  2017         0.000000  Slight or No change
1    MINDTREE     EQ 2017-05-16  ...  2017         0.003631  Slight or No change
2    MINDTREE     EQ 2017-05-17  ...  2017        -0.009681  Slight or No change
3    MINDTREE     EQ 2017-05-18  ...  2017        -0.026563  Slight or No change
4    MINDTREE     EQ 2017-05-19  ...  2017         0.009333  Slight or No change
..        ...    ...        ...  ...   ...              ...                  ...
489  MINDTREE     EQ 2019-05-07  ...  2019        -0.001833  Slight or No change
490  MINDTREE     EQ 2019-05-08  ...  2019         0.000561  Slight or No change
491  MINDTREE     EQ 2019-05-09  ...  2019        -0.000051  Slight or No change
492  MINDTREE     EQ 2019-05-10  ...  2019        -0.005202  Slight or No change
493  MINDTREE     EQ 2019-05-13  ...  2019         0.003845  Slight or No change

[494 rows x 20 columns]

### Query 1.8
Find the average and median values of the column 'Total Traded Quantity' for each of the types of 'Trend'. {Hint : use 'groupby()' on the 'Trend' column and then calculate the average and median values of the column 'Total Traded Quantity'}

In [65]:
mean = df.groupby(df.Trend).mean()['Total Traded Quantity']
median = df.groupby(df.Trend).median()['Total Traded Quantity']
print(f'Required mean is {mean}') 
print(f'Required median is {median}')

Required mean is Trend
Slight or No change    1.388615e+06
Name: Total Traded Quantity, dtype: float64
Required median is Trend
Slight or No change    1006586.5
Name: Total Traded Quantity, dtype: float64


### Query 1.9
SAVE the dataframe with the additional columns computed as a csv file week2.csv. In Module 2, you are going to get familiar with matplotlib, the python module which is used to visualize data.

In [66]:
df.to_csv('week2.csv')